File: src/apps/rustnettests/src/tudp.rs
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use m3::com::Semaphore;
use m3::errors::{Code, Error};
use m3::net::{DGramSocket, DgramSocketArgs, Endpoint, State, UdpSocket, MTU};
use m3::session::NetworkManager;
use m3::test;
use m3::time::TimeDuration;
use m3::vfs::{File, FileEvent, FileRef, FileWaiter};
use m3::{wv_assert_eq, wv_assert_err, wv_assert_ok, wv_run_test};

const TIMEOUT: TimeDuration = TimeDuration::from_secs(1);

pub fn run(t: &mut dyn test::WvTester) {
    // wait once for UDP, because it's connection-less
    wv_assert_ok!(Semaphore::attach("net-udp").unwrap().down());

    wv_run_test!(t, basics);
    wv_run_test!(t, connect);
    wv_run_test!(t, data);
}

fn basics() {
    let nm = wv_assert_ok!(NetworkManager::new("net0"));

    let mut socket = wv_assert_ok!(UdpSocket::new(DgramSocketArgs::new(nm)));

    wv_assert_eq!(socket.state(), State::Closed);
    wv_assert_eq!(socket.local_endpoint(), None);

    wv_assert_ok!(socket.bind(2000));
    wv_assert_eq!(socket.state(), State::Bound);
    wv_assert_eq!(
        socket.local_endpoint(),
        Some(Endpoint::new(crate::NET0_IP.get(), 2000))
    );

    wv_assert_err!(socket.bind(2001), Code::InvState);
}

fn connect() {
    let nm = wv_assert_ok!(NetworkManager::new("net0"));

    let mut socket = wv_assert_ok!(UdpSocket::new(DgramSocketArgs::new(nm)));

    wv_assert_eq!(socket.state(), State::Closed);
    wv_assert_eq!(socket.local_endpoint(), None);

    wv_assert_ok!(socket.connect(Endpoint::new(crate::NET0_IP.get(), 2000)));
    wv_assert_eq!(socket.state(), State::Bound);
}

fn send_recv(
    waiter: &mut FileWaiter,
    socket: &mut FileRef<UdpSocket>,
    dest: Endpoint,
    send_buf: &[u8],
    recv_buf: &mut [u8],
    timeout: TimeDuration,
) -> Result<(usize, Endpoint), Error> {
    wv_assert_ok!(socket.send_to(send_buf, dest));

    waiter.wait_for(timeout);

    if socket.has_data() {
        socket.recv_from(recv_buf)
    }
    else {
        Err(Error::new(Code::Timeout))
    }
}

fn data() {
    let nm = wv_assert_ok!(NetworkManager::new("net0"));

    let mut socket = wv_assert_ok!(UdpSocket::new(DgramSocketArgs::new(nm)));

    wv_assert_ok!(socket.set_blocking(false));

    let dest = Endpoint::new(crate::DST_IP.get(), 1337);

    let mut send_buf = [0u8; 1024];
    for (i, bufi) in send_buf.iter_mut().enumerate() {
        *bufi = i as u8;
    }

    let mut recv_buf = [0u8; 1024];

    let mut waiter = FileWaiter::default();
    waiter.add(socket.fd(), FileEvent::INPUT);

    // do one initial send-receive with a higher timeout than the smoltcp-internal timeout to
    // workaround the high ARP-request delay with the loopback device.
    wv_assert_ok!(send_recv(
        &mut waiter,
        &mut socket,
        dest,
        &send_buf[0..1],
        &mut recv_buf,
        TimeDuration::from_secs(6)
    ));

    wv_assert_err!(
        socket.send_to(&m3::vec![0u8; 4096], dest),
        Code::OutOfBounds
    );
    wv_assert_err!(
        socket.send_to(&m3::vec![0u8; MTU + 1], dest),
        Code::OutOfBounds
    );

    let packet_sizes = [8, 16, 32, 64, 128, 256, 512, 1024];
    for pkt_size in &packet_sizes {
        loop {
            if let Ok((recv_size, src)) = send_recv(
                &mut waiter,
                &mut socket,
                dest,
                &send_buf[0..*pkt_size],
                &mut recv_buf,
                TIMEOUT,
            ) {
                wv_assert_eq!(*pkt_size, recv_size as usize);
                wv_assert_eq!(src, dest);
                wv_assert_eq!(&recv_buf[0..recv_size], &send_buf[0..recv_size]);
                break;
            }
        }
    }
}

File: src/apps/cppnettests/tests/tudp.cc
/*
 * Copyright (C) 2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>

#include <m3/Test.h>
#include <m3/com/Semaphore.h>
#include <m3/net/UdpSocket.h>
#include <m3/session/NetworkManager.h>
#include <m3/vfs/Waiter.h>

#include "../cppnettests.h"

using namespace m3;

static void basics() {
    NetworkManager net("net0");

    auto socket = UdpSocket::create(net);

    WVASSERTEQ(socket->state(), Socket::Closed);
    WVASSERTEQ(socket->local_endpoint(), Endpoint::unspecified());

    socket->bind(2000);
    WVASSERTEQ(socket->state(), Socket::Bound);
    WVASSERTEQ(socket->local_endpoint(), Endpoint(IpAddr(192, 168, 112, 2), 2000));

    WVASSERTERR(Errors::INV_STATE, [&socket] {
        socket->bind(2001);
    });
}

static void connect() {
    NetworkManager net("net0");

    auto socket = UdpSocket::create(net);

    WVASSERTEQ(socket->state(), Socket::Closed);
    WVASSERTEQ(socket->local_endpoint(), Endpoint::unspecified());

    socket->connect(Endpoint(IpAddr(192, 168, 112, 1), 1337));
    WVASSERTEQ(socket->state(), Socket::Bound);
}

static ssize_t send_recv(FileWaiter &waiter, FileRef<UdpSocket> &socket, const Endpoint &dest,
                         const uint8_t *send_buf, size_t sbuf_size, TimeDuration timeout,
                         uint8_t *recv_buf, size_t rbuf_size, Endpoint *src) {
    socket->send_to(send_buf, sbuf_size, dest);

    waiter.wait_for(timeout);

    if(socket->has_data())
        return socket->recv_from(recv_buf, rbuf_size, src);
    return 0;
}

NOINLINE static void data() {
    const TimeDuration TIMEOUT = TimeDuration::from_secs(1);

    NetworkManager net("net0");

    auto socket = UdpSocket::create(net);
    socket->set_blocking(false);

    Endpoint src;
    Endpoint dest = Endpoint(IpAddr(192, 168, 112, 1), 1337);

    uint8_t send_buf[1024];
    for(int i = 0; i < 1024; ++i)
        send_buf[i] = i;

    uint8_t recv_buf[1024];

    FileWaiter waiter;
    waiter.add(socket->fd(), File::INPUT);

    // do one initial send-receive with a higher timeout than the smoltcp-internal timeout to
    // workaround the high ARP-request delay with the loopback device.
    send_recv(waiter, socket, dest, send_buf, 1, TimeDuration::from_secs(6), recv_buf,
              sizeof(recv_buf), &src);

    size_t packet_sizes[] = {8, 16, 32, 64, 128, 256, 512, 1024};

    WVASSERTERR(Errors::OUT_OF_BOUNDS, [&socket, &dest] {
        char tmp[4096];
        static_assert(sizeof(tmp) > NetEventChannel::MAX_PACKET_SIZE, "Packet too small");
        socket->send_to(tmp, sizeof(tmp), dest);
    });
    WVASSERTERR(Errors::OUT_OF_BOUNDS, [&socket, &dest] {
        char tmp[4096];
        socket->send_to(tmp, NetEventChannel::MAX_PACKET_SIZE + 1, dest);
    });

    for(auto pkt_size : packet_sizes) {
        while(true) {
            ssize_t recv_size = send_recv(waiter, socket, dest, send_buf, pkt_size, TIMEOUT,
                                          recv_buf, sizeof(recv_buf), &src);
            if(recv_size != 0) {
                WVASSERTEQ(static_cast<ssize_t>(pkt_size), recv_size);
                WVASSERTEQ(src, dest);

                for(ssize_t i = 0; i < recv_size; ++i)
                    WVASSERTEQ(recv_buf[i], send_buf[i]);
                break;
            }
        }
    }
}

void tudp() {
    // wait for UDP socket just once
    Semaphore::attach("net-udp").down();

    RUN_TEST(basics);
    RUN_TEST(connect);
    RUN_TEST(data);
}
